sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoDomain.java

```diff
@@ -0,0 +1,246 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.protobuf;
+
+import com.google.protobuf.DescriptorProtos;
+import com.google.protobuf.Descriptors;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * ProtoDomain is a container class for Protobuf descriptors. By using a domain for all descriptors
+ * that are related to each other the FileDescriptorSet needs to be serialized only once in the
+ * graph.
+ *
+ * <p>Using a domain also grantees that all Descriptors have object equality, just like statically
+ * compiled Proto classes Descriptors. A lot of Java code isn't used to the new DynamicMessages an
+ * assume always Object equality. Because of this the domain class is immutable.
+ *
+ * <p>ProtoDomains aren't assumed to be used on with normal Message descriptors, only with
+ * DynamicMessage descriptors.
+ */
+public final class ProtoDomain implements Serializable {
+  public static final long serialVersionUID = 1L;
+  private transient DescriptorProtos.FileDescriptorSet fileDescriptorSet;
+  private transient int hashCode;
+
+  private transient Map<String, Descriptors.FileDescriptor> fileDescriptorMap;
+  private transient Map<String, Descriptors.Descriptor> descriptorMap;
+
+  private transient Map<Integer, Descriptors.FieldDescriptor> fileOptionMap;
+  private transient Map<Integer, Descriptors.FieldDescriptor> messageOptionMap;
+  private transient Map<Integer, Descriptors.FieldDescriptor> fieldOptionMap;
+
+  ProtoDomain() {
+    this(DescriptorProtos.FileDescriptorSet.newBuilder().build());
+  }
+
+  private ProtoDomain(DescriptorProtos.FileDescriptorSet fileDescriptorSet) {
+    this.fileDescriptorSet = fileDescriptorSet;
+    hashCode = java.util.Arrays.hashCode(this.fileDescriptorSet.toByteArray());
+    crosswire();
+  }
+
+  private static Map<String, DescriptorProtos.FileDescriptorProto> extractProtoMap(
+      DescriptorProtos.FileDescriptorSet fileDescriptorSet) {
+    HashMap<String, DescriptorProtos.FileDescriptorProto> map = new HashMap<>();
+    fileDescriptorSet.getFileList().forEach(fdp -> map.put(fdp.getName(), fdp));
+    return map;
+  }
+
+  private static Descriptors.FileDescriptor convertToFileDescriptorMap(
```

Comment:
`@Nullable`. I'm a bit sad our analyses didn't catch this.

Reply:
Added @Nullable (stange that github doesn't mark this file as Outdated)